File: linda-logic/src/main/java/it/unibo/coordination/linda/logic/LogicMatch.java
package it.unibo.coordination.linda.logic;

import alice.tuprolog.Struct;
import alice.tuprolog.Term;
import alice.tuprolog.Var;
import it.unibo.coordination.linda.core.Match;
import it.unibo.coordination.prologx.PrologUtils;

import java.util.Map;
import java.util.Optional;

public interface LogicMatch extends Match<LogicTuple, LogicTemplate, String, Term> {

    static LogicMatch failed(LogicTemplate template) {
        return new LogicMatchImpl(template, null, null);
    }

    static LogicMatch wrap(Match<LogicTuple, LogicTemplate, String, Term> match) {
        if (match instanceof LogicMatch) {
            return (LogicMatch) match;
        } else {
            return new LogicMatch() {
                @Override
                public Optional<LogicTuple> getTuple() {
                    return match.getTuple();
                }

                @Override
                public LogicTemplate getTemplate() {
                    return match.getTemplate();
                }

                @Override
                public boolean isMatching() {
                    return match.isMatching();
                }

                @Override
                public Optional<Term> get(String key) {
                    return match.get(key);
                }

                @Override
                public Map<String, Term> toMap() {
                    return match.toMap();
                }

                @Override
                public String toString() {
                    return LogicMatch.toString(this);
                }

                @Override
                public boolean equals(Object o) {
                    return o instanceof LogicMatch && Match.equals(this, (LogicMatch) o);
                }

                @Override
                public int hashCode() {
                    return Match.hashCode(this);
                }

            };
        }
    }

    static Struct getPattern() {
        return Struct.of("match",
                Struct.of("success", Var.of("Success")),
                LogicTemplate.getPattern(),
                Var.of("Tuple"),
                Var.of("Mappings")
        );
    }

    static String toString(LogicMatch match) {
        return match.asTerm().toString();
    }

    default Struct asTerm() {
        return Struct.of("match",
                Struct.of("success", isMatching() ? Struct.atom("yes") : Struct.atom("no")),
                getTemplate().asTerm(),
                getTuple().map(LogicTuple::asTerm).orElse(Struct.atom("empty")),
                Struct.list(
                        toMap().entrySet().stream()
                                .map(kv -> PrologUtils.unificationTerm(kv.getKey(), kv.getValue()))
                )
        );
    }
}

File: linda-logic/src/main/kotlin/it/unibo/coordination/linda/logic/LogicMatch.kt
package it.unibo.coordination.linda.logic

import it.unibo.coordination.linda.core.Match
import it.unibo.tuprolog.core.Scope
import it.unibo.tuprolog.core.Struct
import it.unibo.tuprolog.core.Term
import java.util.*

interface LogicMatch : Match<LogicTuple, LogicTemplate, String, Term> {

    @JvmDefault
    fun asTerm(): Struct = Scope.empty {
        structOf("match",
                structOf("success", if (isMatching) atomOf("yes") else atomOf("no")),
                template.asTerm(),
                tuple.map { it.asTerm() }.orElse(atomOf("empty")),
                listOf(toMap().entries.map { (k, v) -> structOf("=", atomOf(k), v) })
        )
    }

    companion object {

        @JvmStatic
        fun failed(template: LogicTemplate): LogicMatch = LogicMatchImpl(template, null)

        @JvmStatic
        fun wrap(match: Match<LogicTuple, LogicTemplate, String, Term>): LogicMatch =
                if (match is LogicMatch) {
                    match
                } else {
                    object : LogicMatch {
                        override val tuple: Optional<LogicTuple>
                            get() = match.tuple

                        override val template: LogicTemplate
                            get() = match.template

                        override val isMatching: Boolean
                            get() = match.isMatching

                        override fun get(key: String): Optional<Term> {
                            return match[key]
                        }

                        override fun toMap(): Map<String, Term> {
                            return match.toMap()
                        }

                        override fun toString(): String {
                            return toString(this)
                        }

                        override fun equals(other: Any?): Boolean {
                            return other is LogicMatch && Match.equals(this, other as LogicMatch?)
                        }

                        override fun hashCode(): Int {
                            return Match.hashCode(this)
                        }

                    }
                }

        @JvmStatic
        val pattern: Struct
            get() = Scope.empty {
                structOf(
                        "match",
                        structOf("success", varOf("Success")),
                        LogicTemplate.pattern,
                        varOf("Tuple"),
                        varOf("Mappings")
                )
            }

        @JvmStatic
        fun toString(match: LogicMatch): String = match.asTerm().toString()
    }
}
